momentGW/pbc/rpa.py

```diff
@@ -0,0 +1,498 @@
+
+    Parameters
+    ----------
+    gw : BaseKGW
+        GW object.
+    nmom_max : int
+        Maximum moment number to calculate.
+    Lpx : numpy.ndarray
+        Density-fitted ERI tensor, where the first two indices
+        enumerate the k-points (as a dict), the third index is the auxiliary
+        basis function index, and the fourth and fifth indices are
+        the MO and Green's function orbital indices, respectively.
+    integrals : KIntegrals
+        Density-fitted integrals.
+    mo_energy : dict, optional
+        Molecular orbital energies. Keys are "g" and "w" for the Green's
+        function and screened Coulomb interaction, respectively.
+        If `None`, use `gw.mo_energy` for both. Default value is `None`.
+    mo_occ : dict, optional
+        Molecular orbital occupancies. Keys are "g" and "w" for the
+        Green's function and screened Coulomb interaction, respectively.
+        If `None`, use `gw.mo_occ` for both. Default value is `None`.
+    """
+
+    def __init__(
+        self,
+        gw,
+        nmom_max,
+        integrals,
+        mo_energy=None,
+        mo_occ=None,
+    ):
+        """
+        Key quantities for the dRPA calculation. These are constructed in advance to
+        reduce the computational cost of the quadratures.
+
+        returns
+        ----------
+        diag_eri : numpy.ndarray
+            Diagonal of the ERIs for each k-point.
+        d : numpy.ndarray
+            Array of orbital energy differences for each k-point.
+        Liad : dict of numpy.ndarray
+            The (Nkpt, Nkpt)(aux, W occ, W vir) integral array multiplied by the orbital energy
+            differences.
+        Liadinv : dict of numpy.ndarray
+            The (Nkpt, Nkpt)(aux, W occ, W vir) integral array divided by the orbital energy
+            differences.
+        """
+        # TODO discuss with Ollie if this really is optimal or will memory be too much of an issue.
+        super().__init__(gw, nmom_max, integrals, mo_energy, mo_occ)
+
+        kpts = self.kpts
+        Lia = self.integrals.Lia
+
+        self.diag_eri = np.zeros((self.nkpts, self.nkpts), dtype=object)
+        self.d = np.zeros((self.nkpts, self.nkpts), dtype=object)
+
+        self.Liad = np.zeros((self.nkpts, self.nkpts), dtype=object)
+        self.Liadinv = np.zeros((self.nkpts, self.nkpts), dtype=object)
+
+        for q in kpts.loop(1):
+            for kj in kpts.loop(1, mpi=True):
+                kb = kpts.member(kpts.wrap_around(kpts[q] + kpts[kj]))
+                self.diag_eri[q, kb] = np.sum(np.abs(Lia[kj, kb]) ** 2, axis=0) / self.nkpts
+                self.d[q, kb] = util.build_1h1p_energies(
+                    (self.mo_energy_w[kj], self.mo_energy_w[kb]),
+                    (self.mo_occ_w[kj], self.mo_occ_w[kb]),
+                ).ravel()
+
+                self.Liad[q, kb] = Lia[kj, kb] * self.d[q, kb]
+                self.Liadinv[q, kb] = Lia[kj, kb] / self.d[q, kb]
+
+    def integrate(self):
```

Comment:
we can inherit this from the molecular dRPA -- just need to move the quadrature error reporting to a separate function, and redeclare it for k-space

Reply:
With changes to the inheritance of objects like Liad, I don't think this will be possible. Can move them into the integrate class, but then won't be the same as the molecular dRPA.